### ISSUE TITLE
Reproducing CatDQN results

### ISSUE THREAD

**Issue by bmazoure** [astooke/rlpyt#31]
I am trying to reproduce the `AtariCatDQN` results on Atari (testing on Pong for now), with the following parameters:
```
{"agent": {}, "algo": {"discount": 0.99, "batch_size": 128, "learning_rate": 0.00025, "clip_grad_norm": 10.0, "min_steps_learn": 50000, "double_dqn": false, "prioritized_replay": false, "n_step_return": 1, "replay_size": 1000000},
 "env": "pong", "eval_env": {"game": "pong", "episodic_lives": false, "horizon": 27000}, 
"model": {"dueling": false, "n_atoms": 51}, 
"optim": {"eps": 7.8125e-05}, 
"runner": {"n_steps": 50000000.0, "log_interval_steps": 1000000.0}, 
"sampler": {"batch_T": 2, "batch_B": 16, "max_decorrelation_steps": 1000, "eval_n_envs": 4, "eval_max_steps": 125000, "eval_max_trajectories": 100}, 
"name": "async_dqn_pong", "run_ID": 123}
```
but even after a million steps, the average reward is -21.
Here is the `build_and_train` snippet:
```
def build_and_train(args,game="pong", run_ID=0,config=None):
    config['env']['game'] = game
    config["eval_env"]["game"] = config["env"]["game"]
    # Change these inputs to match local machine and desired parallelism.
    affinity = make_affinity(
        # run_slot=0,
        n_cpu_core=8,  # Use 16 cores across all experiments.
        n_gpu=2,  # Use 8 gpus across all experiments.
        gpu_per_run=1,
        sample_gpu_per_run=1,
        async_sample=True,
        optim_sample_share_gpu=True,
        hyperthread_offset=8,
        n_socket=1
        # hyperthread_offset=24,  # If machine has 24 cores.
        # n_socket=2,  # Presume CPU socket affinity to lower/upper half GPUs.
        # gpu_per_run=2,  # How many GPUs to parallelize one run across.
        # cpu_per_run=1,
    )

    sampler = AsyncGpuSampler(
        EnvCls=AtariEnv,
        env_kwargs=config['env'],
        eval_env_kwargs=config['eval_env'],
        **config['sampler']
    )
    algo = CategoricalDQN(
        optim_kwargs=config["optim"], **config["algo"]
    )
    agent = AtariCatDqnAgent(ModelCls=AtariCatDqnModel,model_kwargs=config["model"], **config["agent"])
    runner = AsyncRlEval(
        algo=algo,
        agent=agent,
        sampler=sampler,
        affinity=affinity,
         **config["runner"]
    )
    # config = dict(game=game)
    name = "async_dqn_" + game
    log_dir = "async_dqn"
    with logger_context(log_dir, run_ID, name, config):
        runner.train()
```
I'm not quite sure what the issue is with the setup, as the plots from the arXiv paper seem to suggest training recovers more or less the C51 results.

**Comment by bmazoure** [comment]
To clarify: it does achieve a reward of +20 after 7M steps, but on the arXiv paper convergence is shown to be much faster.

**Comment by astooke** [comment]
Hi, thanks for reaching out!  My first thought is, what is the replayratio (and/or cumreplayratio) which are reported during training?  If this number is very different from 8, then that could affect sample complexity.  This is the main thing which can get out of whack in the async runner.

I think this is the train file used for the CatDQN learning curves: https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/scripts/atari/dqn/train/atari_catdqn_gpu.py

So I would recommend starting with the `MinibatchRlEval` runner, and see if results are good there, before moving onto the `AsyncRlEval` runner.

Otherwise the settings look good!

**Comment by bmazoure** [comment]
Update: with the default setting in the mentionned file, I do manage to reproduce results on Pong and Asteroids.
